### ISSUE TITLE
Need for a Webdriver Copy Script similar to Webdriver clean screenshots

### ISSUE THREAD

**Issue by chnp01** [cerner/terra-toolkit#375]
# Feature Request

## Description
Terra and UX requirement changes impacts team's webdriver automation tests. Whenever any terra component changes or there is a change in UX requirement then these automation tests related screenshots have to be updated. This involves manual effort of copying web driver screenshots from latest to reference based on diff folder.

A utility script can be developed so that the screenshots can be copied from latest folder to reference folder based on diff. This minimizes a lot of manual effort by engineers while testing.

## Additional Context / Screenshots
I have already got the script done and we are currently using it in our project and other team's orion projects. The script I have developed is basic and gets the job done.

## @ Mentions
Spoke with @StephenEsser regarding this functionality and was told to log it as an issue here.

**Comment by emilyrohrbough** [comment]
@chnp01 Could you provide details on how your script works? I could see this being helpful!

**Comment by chnp01** [comment]
@emilyrohrbough  currently its quite basic,
It checks for all the files which are available in the diff folder ( for all locales, viewports, browsers ) and it would execute a move/copy command of those files from latest to reference folder.
The idea being here is that the file/screenshots names are same in all the folders.

This script would copy all the screenshots from latest to reference for all locales based on diff folder. If there are any screenshots that does not need to be copied, then currently, those can be deleted from the diff folder and then the user can execute the script.

**Comment by ryanthemanuel** [comment]
@chnp01 can you link to your script (it's ok if it's internal).

**Comment by nramamurth** [comment]
@chnp01, I'm curious, do you run this script **after** manually validating the differences shown in the 'diff' directory? If not, then I can see the potential of bugs sliding in, in some cases.

**Comment by chnp01** [comment]
> @chnp01, I'm curious, do you run this script **after** manually validating the differences shown in the 'diff' directory? If not, then I can see the potential of bugs sliding in, in some cases.

@nramamurth  yes absolutely. The engineer has to validate it. But most of the cases, these wdio tests fail due to our UX changes or terra upgrades. So, most of the time it would be applicable for all.

This script would copy all the screenshots from latest to reference for all locales based on diff folder. If there are any screenshots that does not need to be copied, then currently, those can be deleted from the diff folder and then the user can execute the script.

**Comment by emilyrohrbough** [comment]
@chnp01 When it copies all of the latest to the reference directly, do you often see diffs generated for screenshots in github that didn't have a diff generated when tests ran? I've noticed that when I've deleted my reference dir and then re-generated all screenshots.

**Comment by StephenEsser** [comment]
I consider this functionality similar to that of the jest update snapshot cli option (and this is how I would anticipate it being used for wdio). Each snapshot/screenshot should still be manually verified before checking into source control.

https://jestjs.io/docs/en/cli#--updatesnapshot

**Comment by chnp01** [comment]
> @chnp01 When it copies all of the latest to the reference directly, do you often see diffs generated for screenshots in github that didn't have a diff generated when tests ran? I've noticed that when I've deleted my reference dir and then re-generated all screenshots.

@emilyrohrbough I guess i might have put it wrong. Let me rephrase it. It would copy the screenshots from latest to reference based on diff folder. So, what ever screenshots are available in the diff folder when we ran wdio are the ones that would be copied from latest

**Comment by ShettyAkarsh** [comment]
yeah, this will be helpfull for the teams when they receive minor version updates for terra. But as @StephenEsser, @nramamurth mentioned solution teams have to manually validate their diff screenshots before running this script.

**Comment by emilyrohrbough** [comment]
Playing round with the script you linked @chnp01, I've gotten to something like:
```js
const fs = require('fs-extra');
const glob = require('glob');

const referenceScreenshots = glob.sync(`${process.cwd()}/**/__snapshots__/reference/**`, { nodir: true, ignore: '**/node_modules/**' });

const updatedScreenshots = [];
const staleScreenshots = [];

Promise.all(referenceScreenshots.map((referenceScreenshot) => {
  const latestScreenshot = referenceScreenshot.replace('reference', 'latest');
  const diffScreenshot = referenceScreenshot.replace('reference', 'diff');

  if (!fs.pathExistsSync(latestScreenshot)) {
    staleScreenshots.push(referenceScreenshot);
  } else if (fs.pathExistsSync(diffScreenshot)) {
    return fs.copy(latestScreenshot, referenceScreenshot, { overwrite: true })
      .then(() => {
        updatedScreenshots.push(referenceScreenshot);
      })
      .catch((err) => {
        console.error(`Error during file copy of ${latestScreenshot}`);
        console.error(err);
      });
  }
  return Promise.resolve();
})).then(() => {
  console.log('Updated Screenshots', updatedScreenshots, '\n\n');
  console.log('Stale Screenshots', staleScreenshots, '\n\n');
}).then(() => {
  const diffDirectories = glob.sync(`${process.cwd()}/**/__snapshots__/diff`, { ignore: '**/node_modules/**' });

  diffDirectories.forEach(diffDir => fs.emptyDirSync(diffDir));
});
```

At this point, I'd be curious to learn more how you run your tests today. Currently if you have screenshot failures, you'd have to copy the latest, then run again to confirm your test run is up to date and passing. Is that how you do it or is your workflow different?

**Comment by chnp01** [comment]
> Playing round with the script you linked @chnp01, I've gotten to something like:
> 
> ```js
> const fs = require('fs-extra');
> const glob = require('glob');
> 
> const referenceScreenshots = glob.sync(`${process.cwd()}/**/__snapshots__/reference/**`, { nodir: true, ignore: '**/node_modules/**' });
> 
> const updatedScreenshots = [];
> const staleScreenshots = [];
> 
> Promise.all(referenceScreenshots.map((referenceScreenshot) => {
>   const latestScreenshot = referenceScreenshot.replace('reference', 'latest');
>   const diffScreenshot = referenceScreenshot.replace('reference', 'diff');
> 
>   if (!fs.pathExistsSync(latestScreenshot)) {
>     staleScreenshots.push(referenceScreenshot);
>   } else if (fs.pathExistsSync(diffScreenshot)) {
>     return fs.copy(latestScreenshot, referenceScreenshot, { overwrite: true })
>       .then(() => {
>         updatedScreenshots.push(referenceScreenshot);
>       })
>       .catch((err) => {
>         console.error(`Error during file copy of ${latestScreenshot}`);
>         console.error(err);
>       });
>   }
>   return Promise.resolve();
> })).then(() => {
>   console.log('Updated Screenshots', updatedScreenshots, '\n\n');
>   console.log('Stale Screenshots', staleScreenshots, '\n\n');
> }).then(() => {
>   const diffDirectories = glob.sync(`${process.cwd()}/**/__snapshots__/diff`, { ignore: '**/node_modules/**' });
> 
>   diffDirectories.forEach(diffDir => fs.emptyDirSync(diffDir));
> });
> ```
> 
> At this point, I'd be curious to learn more how you run your tests today. Currently if you have screenshot failures, you'd have to copy the latest, then run again to confirm your test run is up to date and passing. Is that how you do it or is your workflow different?

When the script wasn't there, yes we used to manually copy the screenshots from latest to reference for the failures and rerun them, but with the script 

Currently, we do it two ways
1. We run wdio in local and validate if the diff s are valid and we execute the copy script, re run wdio to make sure the tests pass and push the changes
2. During Jenkins build, if tests fail, then we would download the artifacts and copy the latest,diff folders into project repo and execute the script and push the changes.
